src/ball.rs

```diff
@@ -0,0 +1,123 @@
+    /// ```
+    ///
+    /// [`Ball`]: struct.Ball.html
+    pub fn new(center: Point<T, D>, radius: T) -> Self {
+        debug_assert!(radius >= T::from_f32(0.0).unwrap());
+        Self { center, radius }
+    }
+
+    /// Returns true if this [`Ball`] contains the [`Point`].
+    ///
+    /// # Examples
+    /// ```
+    /// use bvh::ball::Ball;
+    /// use nalgebra::Point3;
+    ///
+    /// let ball = Ball::new(Point3::new(1.0, 1.0, 1.0), 1.0);
+    /// let point = Point3::new(1.25, 1.25, 1.25);
+    ///
+    /// assert!(ball.contains(&point));
+    /// ```
+    ///
+    /// [`Ball`]: struct.Ball.html
+    pub fn contains(&self, point: &Point<T, D>) -> bool {
+        let mut distance_squared = T::zero();
+        for i in 0..D {
+            distance_squared += (point[i] - self.center[i]).powi(2);
+        }
+        // Squaring the RHS is faster than computing the square root of the LHS.
+        distance_squared <= self.radius.powi(2)
+    }
+
+    /// Returns true if this [`Ball`] intersects the [`Aabb`].
+    ///
+    /// # Examples
+    /// ```
+    /// use bvh::{aabb::Aabb, ball::Ball};
+    /// use nalgebra::Point3;
+    ///
+    /// let ball = Ball::new(Point3::new(1.0, 1.0, 1.0), 1.0);
+    /// let aabb = Aabb::with_bounds(Point3::new(1.25, 1.25, 1.25), Point3::new(3.0, 3.0, 3.0));
+    ///
+    /// assert!(ball.intersects_aabb(&aabb));
+    /// ```
+    ///
+    /// [`Aabb`]: struct.Aabb.html
+    /// [`Ball`]: struct.Ball.html
+    pub fn intersects_aabb(&self, aabb: &Aabb<T, D>) -> bool {
+        // https://gamemath.com/book/geomtests.html#intersection_sphere_aabb
+        // Finding the point in/on the AABB that is closest to the ball's center or,
+        // more specifically, find the squared distance between that point and the
+        // ball's center.
+        let mut distance_squared = T::zero();
+        for i in 0..D {
+            let closest_on_aabb = self.center[i].clamp(aabb.min[i], aabb.max[i]);
+            distance_squared += (closest_on_aabb - self.center[i]).powi(2);
+        }
+
+        // Then test if that point is in/on the ball. Squaring the RHS is faster than computing
+        // the square root of the LHS.
+        distance_squared <= self.radius.powi(2)
+    }
+}
+
+impl<T: BHValue, const D: usize> IntersectsAabb<T, D> for Ball<T, D> {
+    fn intersects_aabb(&self, aabb: &Aabb<T, D>) -> bool {
+        self.intersects_aabb(aabb)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Ball;
+    use crate::testbase::TPoint3;
+
+    #[test]
+    fn ball_contains() {
+        let ball = Ball::new(TPoint3::new(3.0, 4.0, 5.0), 1.5);
+
+        // Ball should contain its own center.
+        assert!(ball.contains(&ball.center));
+
+        // Test some points, selected in 3D software to be just inside/outside the sphere.
+        assert!(ball.contains(&TPoint3::new(3.04605, 3.23758, 3.81607)));
+        assert!(!ball.contains(&TPoint3::new(3.06066, 3.15813, 3.70917)));
```

Comment:
Might be cool to do `let just_inside = TPoint3..` and `let just_outside...` and then compare those? It's a more self-narrating.